libloki/modules/local_loki_server.js

```diff
@@ -79,18 +88,91 @@ class LocalLokiServer extends EventEmitter {
 
     // Start a listening on new server
     return new Promise((res, rej) => {
-      this.server.listen(port, ip, err => {
+      this.server.listen(port, ip, async (err) => {
         if (err) {
           rej(err);
+        } else if (this.upnpClient) {
+          try {
+            const publicPort = await this.punchHole();
+            res(publicPort);
+          } catch (e) {
+            if (e instanceof textsecure.HolePunchingError) {
+              await this.close();
+            }
+            rej(e);
+          }
         } else {
-          res(this.server.address().port);
+          res(port);
         }
       });
     });
   }
 
+  async punchHole() {
+    const privatePort = this.server.address().port;
+    const portStart = 22100;
+    const portEnd = 22200;
+    const ttl = 60 * 15; // renew upnp every 15 minutes
+    const publicPortsInUse = await new Promise((resolve, reject) => {
+      this.upnpClient.getMappings({ local: true }, (err, results) => {
+        if (err) {
+          // We assume an error here means upnp not enabled
+          reject(new textsecure.HolePunchingError('Could not get mapping from upnp. Upnp not available?', err));
+        }
+        else {
+          // remove the current private port from the current mapping
+          // to allow reusing that port.
+          resolve(results
+            .filter(entry => entry.private.port !== privatePort)
+            .map(entry => entry.public.port));
+        }
+      });
+    });
+
+    for (let publicPort = portStart; publicPort <= portEnd; publicPort += 1) {
+      if (publicPortsInUse.includes(publicPort))
```

Comment:
There are a few other if/else without curly's if we are going to enforce it...

Reply:
I wish the linter would take care of that ...
Anyway, fixed!

Reply:
Ya we should see if there is a setting we can add to eslint...
One day™...